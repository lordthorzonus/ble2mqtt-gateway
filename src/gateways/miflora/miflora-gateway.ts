import { GatewayError } from "../ble-gateway";
import { Peripheral } from "@abandonware/noble";
import { DeviceMessage, DeviceType } from "../../types";
import { MiFloraGatewayConfiguration } from "../../config";
import { DeviceRegistryService, handleBleAdvertisement } from "../gateway-helpers";
import { DeviceRegistry, DeviceRegistryEntry } from "../device-registry";
import { MiFloraEventBuffer } from "./miflora-event-buffer";
import { parseMiFloraPeripheralAdvertisement } from "./miflora-parser";
import { transformMiFloraMeasurementsToDeviceMessage } from "./miflora-measurement-transformer";
import { Effect, Option } from "effect";
import { Logger } from "../../infra/logger";

export const mifloraGatewayId = 0x95fe;

export const isMiFloraPeripheral = (peripheral: Peripheral): boolean => {
    const knownMiFloraDeviceNames = ["flower care", "flower mate"];
    const MiFloraMacPrefix = "c4:7c:8d";
    const deviceLocalName = peripheral.advertisement.localName ?? "";

    return (
        knownMiFloraDeviceNames.includes(deviceLocalName.toLowerCase()) ||
        peripheral.address.startsWith(MiFloraMacPrefix)
    );
};

export type ConfiguredMiFloraSensors = MiFloraGatewayConfiguration["devices"];

export const makeMiFloraDeviceRegistry = (
    miFloraSettings: MiFloraGatewayConfiguration,
    globalDefalts: {
        defaultDecimalPrecision: number;
    }
): DeviceRegistry => {
    const deviceSettings = miFloraSettings.devices.map((sensor) => ({
        device: {
            id: sensor.id,
            type: DeviceType.MiFlora,
            friendlyName: sensor.name,
        },
        timeout: sensor.timeout,
    }));

    const defaultDecimalPrecision = miFloraSettings.decimal_precision ?? globalDefalts.defaultDecimalPrecision;

    return new DeviceRegistry(deviceSettings, miFloraSettings.timeout, defaultDecimalPrecision);
};

const handleMifloraBleAdvertisement =
    (miFloraEventBuffer: MiFloraEventBuffer) => (peripheral: Peripheral, deviceRegistryEntry: DeviceRegistryEntry) =>
        Effect.gen(function* () {
            const id = peripheral.uuid;
            const miFloraMeasurementEvent = yield* parseMiFloraPeripheralAdvertisement(peripheral);

            const buffer = miFloraEventBuffer.bufferMeasurementEvent(id, miFloraMeasurementEvent);

            if (miFloraEventBuffer.isBufferReady(buffer)) {
                miFloraEventBuffer.clearBuffer(id);
                return Option.some(
                    transformMiFloraMeasurementsToDeviceMessage(peripheral, deviceRegistryEntry, buffer)
                );
            }

            return Option.none();
        });

export const makeMiFloraGateway = (miFloraSettings: MiFloraGatewayConfiguration) => {
    const sensorEventBuffer = new MiFloraEventBuffer(miFloraSettings.devices);

    return (
        peripheral: Peripheral
    ): Effect.Effect<Iterable<DeviceMessage>, GatewayError, DeviceRegistryService | Logger> => {
        return handleBleAdvertisement(
            peripheral,
            DeviceType.MiFlora,
            false,
            handleMifloraBleAdvertisement(sensorEventBuffer)
        ).pipe(
            Effect.catchTags({
                InvalidMiFloraAdvertisementError: (error) =>
                    new GatewayError({
                        peripheral,
                        message: `Invalid MiFlora advertisement: ${error.message}`,
                    }),
                DeviceNotFoundError: (error) =>
                    new GatewayError({
                        peripheral,
                        message: `Device not found: ${error.id}`,
                    }),
                UnsupportedMiFloraEventError: (error) =>
                    new GatewayError({
                        peripheral,
                        message: `Unsupported MiFlora event: ${error.eventType}`,
                    }),
            })
        );
    };
};
