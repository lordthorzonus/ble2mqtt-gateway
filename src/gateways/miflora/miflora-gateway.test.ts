import { getStubPeripheral } from "./miflora-parser/index.test";
import { MiFloraGateway } from "./miflora-gateway";
import { DateTime, Settings } from "luxon";
import { v4 as uuid } from "uuid";
import { take, toArray } from "rxjs";
import { TestScheduler } from "rxjs/testing";
import { Peripheral } from "@abandonware/noble";

jest.mock("uuid");
const mockedUuid = uuid as jest.Mock;
const defaultTimeout = 30000;

const makeMiFloraGateway = () => {
    return new MiFloraGateway([{ id: "a", name: "flower" }], defaultTimeout);
};

describe("MiFlora Gateway", () => {
    const originalNow = Settings.now;
    const moisturePeripheralAdvertisement = getStubPeripheral("71209800a864aed0a8654c0d08100112", "a");
    const soilConductivityPeripheralAdvertisement = getStubPeripheral("71209800b1cf076e8d7cc40d091002c600", "a");
    const temperaturePeripheralAdvertisement = getStubPeripheral("71209800239c066e8d7cc40d0410020401", "a");
    const illuminancePeripheralAdvertisement = getStubPeripheral("71209800c3cf076e8d7cc40d071003370000", "a");

    describe("handleBleAdvertisement()", () => {
        const testScheduler = new TestScheduler((actual, expected) => {
            expect(actual).toEqual(expected);
        });

        beforeEach(() => {
            Settings.now = () => new Date("2019-10-10T00:00:00.000Z").valueOf();
            Settings.defaultZone = "UTC";
            mockedUuid.mockReturnValue("mock-uuid");
        });

        afterEach(() => {
            Settings.now = originalNow;
            jest.resetAllMocks();
        });

        it.each([
            ["Moisture event", moisturePeripheralAdvertisement],
            ["Soil Conductivity event", soilConductivityPeripheralAdvertisement],
            ["Temperature event", temperaturePeripheralAdvertisement],
            ["Illuminance event", illuminancePeripheralAdvertisement],
        ])(
            "should produce an availability message when receiving %s as the first advertisement from the sensor",
            (_, advertisement) => {
                const gateway = makeMiFloraGateway();

                return new Promise((resolve) => {
                    gateway
                        .handleBleAdvertisement(advertisement)
                        .pipe(toArray())
                        .subscribe((messages) => {
                            expect(messages).toMatchSnapshot();
                            resolve(null);
                        });
                });
            }
        );

        it("should complete without emiting messages if unknown miflora advertisement is received ", (done) => {
            const gateway = makeMiFloraGateway();

            testScheduler.run((helpers) => {
                helpers
                    .expectObservable(
                        gateway.handleBleAdvertisement({
                            ...moisturePeripheralAdvertisement,
                            uuid: "no-such-miflora",
                        } as Peripheral)
                    )
                    .toBe("|");
                done();
            });
        });

        it("should not emit the device message until all sensor events are received once", (done) => {
            const gateway = makeMiFloraGateway();
            testScheduler.run((helpers) => {
                helpers.expectObservable(gateway.handleBleAdvertisement(moisturePeripheralAdvertisement)).toBe("(a|)", {
                    a: expect.objectContaining({ type: "availability" }),
                });
                helpers.expectObservable(gateway.handleBleAdvertisement(temperaturePeripheralAdvertisement)).toBe("|");
                helpers.expectObservable(gateway.handleBleAdvertisement(illuminancePeripheralAdvertisement)).toBe("|");
                helpers.expectObservable(gateway.handleBleAdvertisement(illuminancePeripheralAdvertisement)).toBe("|");
                helpers
                    .expectObservable(gateway.handleBleAdvertisement(soilConductivityPeripheralAdvertisement))
                    .toBe("(a|)", {
                        a: expect.objectContaining({
                            device: expect.objectContaining({ friendlyName: "flower", id: "a" }),
                            type: "sensor-data",
                            id: "mock-uuid",
                            payload: {
                                illuminance: 55,
                                moisture: 18,
                                soilConductivity: 198,
                                temperature: 26,
                            },
                        }),
                    });
                done();
            });
        });
    });
    describe("isMiFloraPeripheral()", () => {});
    describe("observeUnavailableDevices()", () => {
        beforeEach(() => {
            jest.useFakeTimers();
            mockedUuid.mockReturnValue("mock-uuid");
        });

        afterEach(() => {
            jest.useRealTimers();
            jest.resetAllMocks();
        });

        it("should produce unavailability messages for ruuvitags that haven't been observed inside timeout", (done) => {
            const gateway = makeMiFloraGateway();
            gateway
                .handleBleAdvertisement(moisturePeripheralAdvertisement)
                .pipe(toArray())
                .subscribe((messages) => {
                    expect(messages.length).toEqual(1);
                });

            gateway
                .observeUnavailableDevices()
                .pipe(take(1))
                .subscribe((message) => {
                    expect(message).toMatchSnapshot({
                        time: expect.any(DateTime),
                    });
                    done();
                });

            jest.advanceTimersByTime(defaultTimeout + 10000);
        });
    });
});
