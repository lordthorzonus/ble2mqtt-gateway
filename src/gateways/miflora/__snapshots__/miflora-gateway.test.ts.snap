// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MiFlora Gateway handleBleAdvertisement() should produce an availability message when receiving Illuminance event as the first advertisement from the sensor 1`] = `
[
  {
    "device": {
      "friendlyName": "flower",
      "id": "a",
      "macAddress": "aa:bb",
      "rssi": 43,
      "timeout": 30000,
      "type": "miflora",
    },
    "id": "mock-uuid",
    "payload": {
      "state": "online",
    },
    "time": "2019-10-10T00:00:00.000Z",
    "type": "availability",
  },
]
`;

exports[`MiFlora Gateway handleBleAdvertisement() should produce an availability message when receiving Moisture event as the first advertisement from the sensor 1`] = `
[
  {
    "device": {
      "friendlyName": "flower",
      "id": "a",
      "macAddress": "aa:bb",
      "rssi": 43,
      "timeout": 30000,
      "type": "miflora",
    },
    "id": "mock-uuid",
    "payload": {
      "state": "online",
    },
    "time": "2019-10-10T00:00:00.000Z",
    "type": "availability",
  },
]
`;

exports[`MiFlora Gateway handleBleAdvertisement() should produce an availability message when receiving Soil Conductivity event as the first advertisement from the sensor 1`] = `
[
  {
    "device": {
      "friendlyName": "flower",
      "id": "a",
      "macAddress": "aa:bb",
      "rssi": 43,
      "timeout": 30000,
      "type": "miflora",
    },
    "id": "mock-uuid",
    "payload": {
      "state": "online",
    },
    "time": "2019-10-10T00:00:00.000Z",
    "type": "availability",
  },
]
`;

exports[`MiFlora Gateway handleBleAdvertisement() should produce an availability message when receiving Temperature event as the first advertisement from the sensor 1`] = `
[
  {
    "device": {
      "friendlyName": "flower",
      "id": "a",
      "macAddress": "aa:bb",
      "rssi": 43,
      "timeout": 30000,
      "type": "miflora",
    },
    "id": "mock-uuid",
    "payload": {
      "state": "online",
    },
    "time": "2019-10-10T00:00:00.000Z",
    "type": "availability",
  },
]
`;
